### ISSUE TITLE
Not able to load keys/certs using pcks#11 URI and executable files

### ISSUE THREAD

**Issue by Dung-Han-Lee** [OpenSC/libp11#367]
Hello, 
I am trying to establish TLS connection using hardware secure modules (smart card) on a x86, ubuntu 16.04 machine. I am able to load keys using openssl engine then use it to launch s_client:

> engine -t -pre SO_PATH:/usr/local/lib/engines-1.1/libpkcs11.so -pre ID:pkcs11 -pre LIST_ADD:1 -pre LOAD -pre MODULE_PATH:/usr/lib/opensc-pkcs11.so
> 
> openssl s_client -engine pkcs11 -connect localhost:10000 -cert client.crt -key 1:01 -keyform engine

However, I couldn't get the engine code working, after compiling the check-privatekey.c under /tests folder, and run it with the following command:

> ./check-privatekey.o "pkcs11:token=YubiHSM;id=%01;type=public" "pkcs11:token=YubiHSM;id=%01;type=private" /usr/lib/opensc-pkcs11.so engines.cnf.in

I receive the following error:

> PKCS#11: Initializing the engine
> Found 2 slots
> Looking in slot -1 for certificate: id=01
> [18446744073709551615] Virtual hotplug slot       no tok          
> [1] Yubico Yubikey 4 OTP+U2F+  login             (PIV_II (PIV Card Holder pin))
> Specified object not found
> Looking in slot -1 for certificate: id=01
> [18446744073709551615] Virtual hotplug slot       no tok          
> [1] Yubico Yubikey 4 OTP+U2F+  login             (PIV_II (PIV Card Holder pin))
> Specified object not found
> Could not get certificate pkcs11:token=YubiHSM;id=%01;type=public

I am not sure if I am approaching this issue correctly, I choose to compile check-privatekey.c because it involves openssl engine controls that I think would be required for my usage. I can't find example online about my specific use case, nor could I find documentation on this repo. Is there any documentation on this? If not, could someone suggest how could I replicate the aforementioned openssl command lines in .c files? Thank you in advance.

**Comment by mouse07410** [comment]
Yubikey and YubiHSM are very different. 

To access keys on Yubikey, remove "token=YubiHSM;" part.

To access keys on YubiHSM, change _opensc-pkcs11.so_ to the one provided by YubiHSM-2 SDK.

**Comment by Dung-Han-Lee** [comment]
Hello Mouse,
Thank you for your quick reply, I will look into this tomorrow. One thing to clarify: I assume what could be achieved with openssl command lines can be replicated using c code. Since I got it working with opensc-pkcs11.so, I assume the same would work for c code as well?

**Comment by Dung-Han-Lee** [comment]
Hello @mouse07410, I followed your advice and removed the "token=YubiHSM" since I am using a Yubikey.

> ./check-privatekey.o "pkcs11:id=%01;type=public" "pkcs11:id=%01;type=private" /usr/lib/opensc-pkcs11.so engines.cnf.in

It seems that the certificate is successfully loaded. However, I am not able to load the private key. There are two questions associated with the output:

1. The console prompt me to input pin, but I did not set up pin for my yubikey yet. I did not observe the same behavior when using openssl command lines though. Am I missing out something? Can I expect openssl command line behave the same as C code?
2. It seems that the program make several attempts to the load the key despite there's only 2 slots found. Is there any explanation on this?

PKCS#11: Initializing the engine
Found 2 slots
Looking in slot -1 for certificate: id=01
[18446744073709551615] Virtual hotplug slot       no tok          
[1] Yubico Yubikey 4 OTP+U2F+  login             (PIV_II (PIV Card Holder pin))
Found slot:  Yubico Yubikey 4 OTP+U2F+CCID 00 00
Found token: PIV_II (PIV Card Holder pin)
Found 1 cert:
Loading private key "pkcs11:id=%01;type=private"
Looking in slot -1 for key: id=01
[18446744073709551615] Virtual hotplug slot       no tok          
[1] Yubico Yubikey 4 OTP+U2F+  login             (PIV_II (PIV Card Holder pin))
Found slot:  Yubico Yubikey 4 OTP+U2F+CCID 00 00
Found token: PIV_II (PIV Card Holder pin)
Loading private key "pkcs11:id=%01;type=private"
Looking in slot -1 for key: id=01
[18446744073709551615] Virtual hotplug slot       no tok          
[1] Yubico Yubikey 4 OTP+U2F+  login             (PIV_II (PIV Card Holder pin))
Found slot:  Yubico Yubikey 4 OTP+U2F+CCID 00 00
Found token: PIV_II (PIV Card Holder pin)
Enter PKCS#11 token PIN for PIV_II (PIV Card Holder pin):
UI_process failed
No PIN code was entered
Login to token failed, returning NULL...
PKCS11_get_private_key returned NULL
Could not load key
At main.c:157:
- SSL error:28078065:UI routines:UI_set_result_ex:result too small: crypto/ui/ui_lib.c:905
- SSL error:2807106B:UI routines:UI_process:processing error: crypto/ui/ui_lib.c:545
- SSL error:26096080:engine routines:ENGINE_load_private_key:failed loading private key: cr
